### ISSUE TITLE
Engine abstraction

### ISSUE THREAD

**Issue by brson** [tikv/tikv#6402]
This issue is to organize and  track progress on abstracting TiKV over generic storage
engines (previous issue: https://github.com/tikv/tikv/issues/4184). There are many tasks to be done and contributions are welcome.

In scope: the process of abstracting the entire RocksDB engine API into the `engine_traits` and `engine_rocks` crates, decoupling TiKV and its test suite from RocksDB., and prototyping new engine bindings. Out of scope: refactoring `engine_traits` to better support with non-RocksDB engines.

The [`engine_traits` crate docs][engine_traits] describe the design and
migration process.

The issue is updated to contain a checklist of tasks that definitely need to
happen, along with name of who is working on it. When you decide to take a task,
say so on the issue before attempting it so that someone else doesn't duplicate
your work.

It's hard to know _in what order_ these need to happen, nor exactly how to
complete each of them, as each tends to present new challenges. Instead I have
documented here, in [the `engine_traits` crate docs][engine_traits], and in a [status report](todo), my experiences so far and suggestions for how to proceed.

[engine_traits]: https://github.com/tikv/tikv/blob/8b21c0f364b13ee3452f1c6c4d0526169e3d6481/components/engine_traits/src/lib.rs

The most important thing to keep in mind is:

Make small changes: small commits, small PRs. When you find that something else
needs to change first, stop what you are doing, and make that change, and submit
that PR. Don't let your patches grow out of control. Don't be afraid to throw away
your work and try something else.

### Migrating the `engine` abstractions

Take some abstraction defined in `engine`, implement it in `engine_traits` and
`engine_rocks`, change all callers to use the new version, delete the old
version. This abstraction can be a single function, an implementation of a
trait, a type, or an entire trait.

- [x] SstReader/Writer (brson)
- [x] IOLimiter (brson)
- [x] Snapshot (brson)
- [ ] Mutable for WriteBatch
- [ ] Peekable for DB
- [ ] Iterable for DB (hicqu)
- [ ] cf module
- [ ] util module
- [ ] rocks/util module
- [ ] Errors
- [ ] Engines

### Eliminating direct-use of `rocksdb` re-exports

`engine` reexports many types from `rocksdb`. Find one of those types, add an
equivalent abstraction to `engine_traits`, implement it for `engine_rocks`,
migrate all callers to use `engine_rocks`, delete the reexport so that it cannot
be used again.

- [ ] TablePropertiesCollection (brson)
- [ ] TablePropertiesCollector (brson)
- [ ] TablePropertiesCollectorFactory (brson)
- [ ] BlockBasedOptions
- [ ] CColumnFamilyDescriptor
- [ ] CFHandle
- [ ] Cache
- [ ] ColumnFamilyOptions
- [ ] CompactOptions
- [ ] CompactionJobInfo
- [ ] CompactionOptions
- [ ] CompactionPriority
- [ ] DB
- [ ] DBBackgroundErrorReason
- [ ] DBBottommostLevelCompaction
- [ ] DBCompactionStyle
- [ ] DBCompressionType
- [ ] DBEntryType
- [ ] DBIterator
- [ ] DBOptions
- [ ] DBRateLimiterMode
- [ ] DBRecoveryMode
- [ ] DBStatisticsHistogramType
- [ ] DBStatisticsTickerType
- [ ] DBTitanDBBlobRunMode
- [ ] DBVector
- [ ] Env
- [ ] EnvOptions
- [ ] EventListener
- [ ] ExternalSstFileInfo
- [ ] FlushJobInfo
- [ ] HistogramData
- [ ] IngestExternalFileOptions
- [ ] IngestionInfo
- [ ] Kv
- [ ] LRUCacheOptions
- [ ] MapProperty
- [ ] MemoryAllocator
- [ ] PerfContext
- [ ] Range
- [ ] RateLimiter
- [ ] ReadOptions
- [ ] SeekKey
- [ ] SequentialFile
- [ ] SliceTransform
- [ ] TitanBlobIndex
- [ ] TitanDBOptions
- [ ] UserCollectedProperties
- [ ] Writable
- [ ] WriteBatch
- [ ] WriteOptions
- [ ] WriteStallCondition
- [ ] WriteStallInfo
- [ ] Writebatch
- [ ] load_latest_options,
- [ ] supported_compression,
- [ ] run_ldb_tool,
- [ ] set_external_sst_file_global_seq_no,

### "Pulling up" the generic abstractions though TiKV

Find the concrete users of these types and make them generic over the
appropriate traits.

When attempting this try to go in small increments: change one instance of
RocksFoo to the generic "E: Foo", and change its caller to introduce the
concrete RocksFoo, get it to compile, then repeat.

- [x] RocksSstReader/Writer (brson)
- [x] RocksIOLimiter (brson)
- [ ] RocksSnapshot (brson)
- [ ] RocksWriteBatch
- [ ] RocksEngine
- TODO what else?

### Isolating test cases from RocksDB

- [ ] Create an `engine_test` crate that simply wraps `engine_rocks`. Don't use
  typedefs. Actually define wrapping types. This creates a layer of indirection
  between TiKV test cases and RocksDB so that TiKV can be fully-decoupled from
  RocksDB.
- [ ] Migrate test module X to `engine_test` (TODO expand this)
- [ ] Write a `engine_traits` test suite that is parameterized over implementations

### Misc

- [ ] Create an `engine_panic` crate that contains no behavior but may be used
  as a template for creating new engines, avoiding the time it takes to redifine
  all the impls required by an engine and keeping engine organization
  consistent. (brson)

**Comment by brson** [comment]
It's kinda buried in the op, but for anybody wishing to contribute, please read my experience report about what we've done so far and what I've learned in the process: https://gist.github.com/f6aa58bebc7de16ea17741bce31f6080.

**Comment by brson** [comment]
FWIW I've been working on-and-off on abstracting TablePropertiesCollection and its associated types for about a month, trying different techniques. I'm now writing a completely new set of bindings within rust-rocksdb that use reference counting.

**Comment by Rustin-Liu** [comment]
I am working on  rocks/util module.

**Comment by brson** [comment]
Thanks @Rustin-Liu !

**Comment by brson** [comment]
Per my comment on https://github.com/tikv/tikv/pull/6661#issuecomment-589486398, it looks like after @Rustin-Liu's migration of the `MetricsFlusher` there will still be more work to be done on it.

MetricsFlusher needs to be implemented in the engine_traits crate so that it can be instantiated with two different engines (vs. being a trait that is implemented in each engine). Doing so will require refactoring the implementation and abstracting out some of the calls it makes, as well as the lists of metrics it uses.

**Comment by brson** [comment]
IOLimiter no longer exists, replaced by async-speed-limit in 965e8bfbca1aa1043a4785a419972bbe9d5c957c

**Comment by brson** [comment]
Now that raftstore is its own crate (thanks @overvenus), we have a nice intermediate goal: remove the `engine` dependency from `raftstore`. That simplifies the problem quite a bit.

**Comment by brson** [comment]
@Rustin-Liu a good next util to abstract would be `util::compact_range`. This is something preventing us from fully abstracting raftstore over engines.

It looks like that will require some new abstractions in engine_traits, perhaps a new file with a CompactExt trait that contains compact_range_cf_opt. I have not looked closely though.

**Comment by brson** [comment]
I've updated op with current progress, and added a task to document engine_traits cc @hicqu

**Comment by brson** [comment]
Per https://github.com/tikv/tikv/pull/7371/files/6498ad5f98fcf2fd12ffbf864b02f2e88ee12d68#r405944139, there are a number of types in raftstore that are passing around `E::Snapshot` and propagating ugly type parameters.

`Snapshot` itself is a very complex type that I don't see any hope in making into a dynamic object, but it looks to me like this particular snapshot is used only for computing a hash. We might be able to create a new object-safe type that can package up the snapshot and do nothing but compute that hash in a boxable way.

**Comment by Rustin-Liu** [comment]
Hi @brson, I want to working on `Remove protobuf dependency `. 
Could you explain why we need to remove this dependency and add more detail about that?

**Comment by brson** [comment]
@Rustin-Liu 

re "remove protobuf dependency", the Mutable and SyncMutable traits have some convenience methods (put_msg, put_msg_cf) for writing protobuf messages. There's no fundamental reason these need to be conflated with the engine itself, and I think should live at a higher level of abstraction from the engine traits.

The implementations of these methods should live outside of engine_traits, probably in tikv itself, or if necessary, some helper crate.

**Comment by brson** [comment]
@5kbpers has implemented an in-memory engine in https://github.com/tikv/tikv/pull/8430

**Comment by brson** [comment]
This has made a lot of progress. We're still months away from being able to run tikv with a real alternative engine, but most of the pieces are mostly in place now.

- The raftstore crate is almost completely parameterized over the KvEngine trait, but is not yet free of the engine_rocks dependency. There is perhaps one more test still using RocksEngine directly, and the crate still uses the engine_rocks::PerfContext global state.
- The engine_test crate provides a mechanism for picking which engine to construct at compile time, and is used in the raftstore crate to make it testable with other engines, currently just engine_rocks and engine_panic

There's enough scaffolding to begin adding other engines, and a few are in progress:

- @5kbpers is working on a skiplist-based in-memory engine
- @skyzh is working on https://github.com/tikv/agatedb
- I am working on a simple engine while I build an engine conformance test suite, engine_traits_tests

The way to implement an engine is:

- Copy engine_panic to engine_myengine
- Make it a dependency of engine_test, and fill in the boilerplate in engine_test/src/lib.rs so that it can be constructed for test cases
- Add engine_myengine to tikv/Cargo.toml and fill out feature flags, copying from engine_panic

Then you should be able to run tests against raftstore with

```
cargo test -p raftstore --no-default-features --features=protobuf-codec,test-engines-myengine
```

Currently I am working on:

- Creating a conformance test suite for engines, called engine_traits_tests. This should help a lot when adding new engines.
- Moving engine_rocks dependencies from the tikv crate

**Comment by brson** [comment]
Here's an issue about dealing with the engine configuration: https://github.com/tikv/tikv/issues/9242

**Comment by brson** [comment]
I've updated the op to drastically simplify the remaining task list

**Comment by hbina** [comment]
So you guys are at stage 3 now? Or how has this changed.

**Comment by brson** [comment]
I've updated the op to indicate completion of a few items.

@hbina Based on that link, which I wrote pretty long ago now, I'd say we're in stage 3, and have been for a while.

The way I've been going about it is to extract small crates from the tikv crate that depend on rocksdb, then refactor that extracted crate to depend on engine_traits but not engine_rocks.

We've completed the raftstore crate, and now are working on the tikv_kv crate.

Stage "4", about making the test suite use rocksdb is underway - there's an engine_test crate that should be used in place of the concrete engine_rocks crate, and is used to test raftstore.

We also have a partial engine-only test suite in tree, engine_traits_tests, and I have bindings to a personal toy database that passes that test suite.

I have also begun bindings to https://github.com/tikv/agatedb.

**Comment by brson** [comment]
I guess I should point out that we are far enough along that anybody that wants to can _begin_ prototyping engine_traits implementations for alternate engines, now that we have engine_traits_tests merged. I myself have begun two out of tree.

Getting a new engine to pass engine_traits_tests is the first step to adding a new engine.

Rough instructions are [in this comment](https://github.com/tikv/tikv/issues/6402#issuecomment-714802531)

An engine can be tested against the engine tests with e.g.:

```
cargo test -p engine_traits_tests --no-default-features --features=protobuf-codec,test-engines-agatedb
```

It will still be many months before tikv can fully run on other engines. There are lots of tiny problems to solve yet, and some of the abstractions exported by engine_traits are two rocks-specific to remain.

**Comment by brson** [comment]
I've posted a new issue to remove the table properties abstractions from engine traits: https://github.com/tikv/tikv/issues/10068

**Comment by brson** [comment]
This PR finally makes it possible to launch tikv-server with a non-rocks engine: https://github.com/tikv/tikv/pull/10529

**Comment by brson** [comment]
Status update:

TiKV is almost able to build and even run with engines other than rocks, but the final PR to enable other engines, https://github.com/tikv/tikv/pull/10945, was reverted due to hitting an unwrap on some tests that I am not familiar with. It should be relatively easy to revive that branch and fix the unwrap issue, but I have not investigated. Once that branch is merged, https://github.com/tikv/tikv/issues/10528 can be closed.

Building on that patch, I have already proceeded to demonstrate tikv running another engine, the most recent attempt being my engine-simple-try-2 branch, https://github.com/brson/tikv/tree/engine-simple-try-2. This branch was at the time able to launch tikv with my own engine built specifically for testing the engine traits and reach a steady state where it could do simple writes and reads.

The engine I developed for this purpose is at https://github.com/brson/blocksy3. This engine is designed for easy readability, and to support exactly the features required by TiKV, but not for performance.

TiKV contains a set of tests specifically designed for bootstrapping new engines, engine_traits_tests, the usage of which is documented in that crate. An engine that passes that test suite should more-or-less be usable within TiKV for the simplest use cases, at which point further work can be put into filling out test cases for less-used parts of the engine traits API.

TiKV is not fully abstracted over engines - there are a few subsystems which are dynamically disabled for non-rocks engines, but which should be relatively easy to rediscover once somebody digs in, and which are partially described here https://github.com/tikv/tikv/pull/10529

An issue for creating agatedb bindings is open here https://github.com/tikv/tikv/issues/9844

I expect the design of the engines traits to change significantly as support for other engines is added. The current design is just a thin layer over rocks, with only the most obviously unimplementable parts of rocks papered-over with very high level abstractions.

**Comment by skyzh** [comment]
Good to know that! Hope that we can integrate more engines into TiKV.

**Comment by cscetbon** [comment]
Hey @brson any news on this ?
